### ISSUE TITLE
torch.compile mode='reduce-overhead' regression: 7x slower cudaGraphLaunch and new cudagraph_trees.py overhead in PyTorch 2.10 vs 2.9

### ISSUE THREAD

**Issue by Heronalps** [pytorch/pytorch#174575]
## 🐛 Describe the bug

`torch.compile(mode="reduce-overhead")` has a **40-55% throughput regression** in PyTorch 2.10 vs 2.9 on an H100 GPU running a 6B parameter production GPT inference engine. The root cause is **7x slower `cudaGraphLaunch`** calls and **new Python-level overhead** in `cudagraph_trees.py` that does not exist in 2.9.

We performed a controlled isolation experiment with three configurations to separate the PyTorch code regression from the CUDA toolkit change, and found **both PyTorch 2.10's inductor code and the CUDA 13.x toolkit contribute**, with PyTorch 2.10's code changes being the primary factor.

## Environment

- **GPU**: NVIDIA H100 80GB HBM3 (Modal cloud)
- **PyTorch versions tested**: 2.9.0+cu128, 2.10.0+cu128, 2.10.0+cu130
- **CUDA toolkits tested**: 12.8, 13.0/13.1
- **OS**: Ubuntu 24.04 (container)
- **Python**: 3.10
- **Workload**: 6B parameter GPT model, `torch.compile(fullgraph=True, mode="reduce-overhead")`, 160 concurrent requests, batched inference

### Three-way isolation experiment (4 rounds x 500 steps, round 0 discarded, 3 scored)

| Configuration | Med Tok/s | StdDev | CV | Regression |
|---|---|---|---|---|
| **torch 2.9.0+cu128** (CUDA 12.8) | **8,016** | 160 | 2.0% | baseline |
| **torch 2.10.0+cu128** (CUDA 12.8) | **6,639** | 474 | 7.1% | **-17.2%** |
| **torch 2.10.0+cu130** (CUDA 13.0) | **4,792** | 233 | 4.9% | **-40.2%** |

The isolation experiment (same CUDA 12.8, different PyTorch) confirms PyTorch 2.10's code is responsible for at least 17% throughput loss independent of CUDA version.

### Per-round stability

| Config | R0 (discard) | R1 | R2 | R3 |
|---|---|---|---|---|
| torch 2.9 + CUDA 12.8 | 8,122 | 7,873 | 8,189 | 7,984 |
| torch 2.10 + CUDA 12.8 | 7,782 | 7,129 | 6,606 | 6,182 |
| torch 2.10 + CUDA 13.1 | 5,313 | 5,046 | 4,743 | 4,588 |

**Notable**: Both torch 2.10 configs show **round-over-round degradation** (declining throughput over sequential measurement windows) that torch 2.9 does not exhibit. This suggests state accumulation in the CUDA graph tree management code.

## Root Cause: `cudaGraphLaunch` + `cudagraph_trees.py`

Perfetto traces captured with `torch.profiler` show:

### 1. `cudaGraphLaunch` is 7x slower per call

| Metric | Torch 2.9 | Torch 2.10 | Change |
|---|---|---|---|
| `cudaGraphLaunch` total | 41.25 ms | 293.27 ms | **+611%** |
| `cudaGraphLaunch` per call | 0.84 ms | 5.99 ms | **+7.1x** |
| Launch count | 49 | 49 | Same |
| Graph instantiations | 1 | 1 | Same |

All 49 launches shifted from the [0.5-1.0) ms range (96% in 2.9) to the [5.0-10.0) ms range (100% in 2.10).

### 2. New Python overhead in `cudagraph_trees.py`

| Component | Torch 2.9 | Torch 2.10 | Change |
|---|---|---|---|
| `cudagraph_tree` total | 758 ms | 2,866 ms | **+278%** |
| `cudagraph_tree` avg/event | 0.124 ms | 0.471 ms | **+280%** |
| `CompiledFxGraph` avg | 4.41 ms | 7.02 ms | +59% |

### 3. Functions that appear in 2.10 but NOT in 2.9 traces

| Function | Count | Total Time |
|---|---|---|
| `cudagraph_trees.py(2029)` | 97 | 667.76 ms |
| `cudagraph_trees.py(2094)` | 97 | 667.36 ms |

These are entirely new overhead that did not exist in the 2.9 code path.

## What is NOT the cause

- **Graph structure**: Same number of graphs (1), same launch count (49), same instantiation count (1)
- **GPU kernels**: The actual CUDA kernels (GEMM, attention, triton fusions) are slightly *faster* in 2.10
- **CUDA toolkit alone**: Keeping CUDA 12.8 while upgrading to torch 2.10 still shows 17% regression
- **Compile mode**: Tried `max-autotune`, `default`, `no-cudagraph`, `no-compile` — none recover performance. The `reduce-overhead` baseline remains the best strategy for both versions.

## Reproducer

The workload is a production GPT model, but the issue should be reproducible with any model using `torch.compile(mode="reduce-overhead")` with CUDA graphs. The key characteristics:

```python
import torch

model = ...  # any model
compiled_model = torch.compile(model, fullgraph=True, mode="reduce-overhead")

# Warm up (captures CUDA graph)
for _ in range(10):
    output = compiled_model(input)

# Measure — this is where the regression shows
# Each step's cudaGraphLaunch takes 7x longer in 2.10 vs 2.9
for _ in range(500):
    torch.cuda.synchronize()
    t0 = time.perf_counter()
    output = compiled_model(input)
    torch.cuda.synchronize()
    elapsed = time.perf_counter() - t0
    # 2.9: ~12ms per step, 2.10: ~28ms per step (on our 6B model)
```

## Expected behavior

`cudaGraphLaunch` should take roughly the same time per call in 2.10 as in 2.9 (~0.84ms), and the new `cudagraph_trees.py` code paths should not add 667ms of Python overhead per 50 inference steps.

## Additional context

- The round-over-round degradation pattern (throughput declining across sequential measurement rounds without container restart) in 2.10 but not 2.9 suggests the CUDA graph tree management may be accumulating state or doing increasing work per replay
- We have full Perfetto traces for both versions available on request
- Our production system serves 1M+ DAU and we are blocked from upgrading to 2.10 due to this regression

cc @jerryzh168 @mcarilli @ezyang @eellison @penguinwu @BoyuanFeng @chauhang @pytorchbot

**Comment by ezyang** [comment]
There's an integration point to swap in a faster cuda graph wrapper; vllm uses it. Check it out.

**Comment by ngimel** [comment]
@Cheng-Logic you didn't conduct a logical audit on this throughput collapse, claude code did, and as such your comment violates our contributing policy (see https://github.com/albanD/pytorch/blob/f971ec007860618dd2e9478fd775f385d44bd194/CONTRIBUTING.md#ai-assisted-development). None of the copy-paste you did even remotely explains why cudaGraphLaunch would become slower, and cudagraph_trees didn't have any substantive changes between 2.9 and 2.10 https://github.com/pytorch/pytorch/commits/main/torch/_inductor/cudagraph_trees.py, what would be the exact commit you would refer to as introducing extra overhead? No extra pytho-side validation would have any effect on cudaGraphLaunch duration.
